kernel/sched.c

```diff
@@ -1564,6 +1566,10 @@ static int32_t z_tick_sleep(k_ticks_t ticks)
 	if (ticks > 0) {
 		return ticks;
 	}
+#else
+	/* busy wait to be time coherent since subsystems may depend on it */
+	while (expected_wakeup_ticks > sys_clock_tick_get_32())
```

Comment:
Are we actually guaranteed to have a timer driver when MULTITHREADING=n?  The original idea was to eliminate everything, including things like timing.  There's no spec for this feature, so... maybe.  But IMHO bloating the system for just this case seems like a bad trade?

You might also consider k_busy_wait(), which is "usually" based on the same system timer, but doesn't have to be.  On Nordic, it's a CPU-counted loop and thus won't require the driver.

Reply:
Hi @andyross ,

> Are we actually guaranteed to have a timer driver when MULTITHREADING=n? The original idea was to eliminate everything, including things like timing.

I saw that `sys_clock_tick_get_32()` returns 0 when no `CONFIG_SYS_CLOCK_EXISTS`. Otherwise it is well even for `CONFIG_TICKLESS_KERNEL=y`. Besides that I agreed that seems better to use the `k_busy_wait()` since it is there.